src/FSharp.Data.GraphQL.Server.AppInfrastructure/GraphQLWebsocketMiddleware.fs

```diff
@@ -0,0 +1,375 @@
+    let observer = new Reactive.AnonymousObserver<Output>(
+      onNext =
+        (fun theOutput ->
+          theOutput
+          |> howToSendDataOnNext id
+          |> Task.WaitAll
+        ),
+      onError =
+        (fun ex ->
+          logger.LogError("[Error on subscription {id}]: {exceptionstr}", id, ex)
+        ),
+      onCompleted =
+        (fun () ->
+          Complete id
+          |> sendMessageViaSocket jsonSerializerOptions (socket)
+          |> Async.AwaitTask
+          |> Async.RunSynchronously
+          subscriptions
+          |> GraphQLSubscriptionsManagement.removeSubscription(id)
+        )
+    )
+
+    let unsubscriber = streamSource.Subscribe(observer)
+
+    subscriptions
+    |> GraphQLSubscriptionsManagement.addSubscription(id, unsubscriber, (fun _ -> ()))
+
+  let tryToGracefullyCloseSocket (code, message) theSocket =
+    task {
+        if theSocket |> canCloseSocket
+        then
+          do! theSocket.CloseAsync(code, message, new CancellationToken())
+        else
+          ()
+    }
+
+  let tryToGracefullyCloseSocketWithDefaultBehavior =
+    tryToGracefullyCloseSocket (WebSocketCloseStatus.NormalClosure, "Normal Closure")
+
+  let handleMessages (cancellationToken: CancellationToken) (serializerOptions: JsonSerializerOptions) (executor : Executor<'Root>) (root: unit -> 'Root) (pingHandler : PingHandler option) (socket : WebSocket) =
+    let subscriptions = new Dictionary<SubscriptionId, SubscriptionUnsubscriber * OnUnsubscribeAction>()
+    // ---------->
+    // Helpers -->
+    // ---------->
+    let safe_ReceiveMessageViaSocket = receiveMessageViaSocket (new CancellationToken())
+
+    let safe_Send = sendMessageViaSocket serializerOptions socket
+    let safe_Receive() =
+      socket
+      |> safe_ReceiveMessageViaSocket serializerOptions executor
+
+    let safe_SendQueryOutput id output =
+      let outputAsDict = output :> IDictionary<string, obj>
+      match outputAsDict.TryGetValue("errors") with
+      | true, theValue ->
+        // The specification says: "This message terminates the operation and no further messages will be sent."
+        subscriptions
+        |> GraphQLSubscriptionsManagement.removeSubscription(id)
+        safe_Send (Error (id, unbox theValue))
+      | false, _ ->
+        safe_Send (Next (id, output))
+
+    let sendQueryOutputDelayedBy (cancToken: CancellationToken) (ms: int) id output =
+      task {
+            do! Async.StartAsTask(Async.Sleep ms, cancellationToken = cancToken)
+            do! output
+                |> safe_SendQueryOutput id
+        }
+    let safe_SendQueryOutputDelayedBy = sendQueryOutputDelayedBy cancellationToken
+
+    let safe_ApplyPlanExecutionResult (id: SubscriptionId) (socket) (executionResult: GQLResponse)  =
```

Comment:
Is the `safe_` prefix here used to indicate the opposite of an unsafe action? If it's the case, I would suggest to just remove this prefix. I think it's better to name unsafe methods with `unsafe` prefixes or suffixes instead.

Reply:
Great catch!

Reply:
This was initially supposed to mean that it was being executed with a cancellation token that was already there (no need to provide one again). I agree that `safe` is not the best name choice here and I just used it for lack of inspiration for a better name. On the other hand, the functions that do take a cancellation token are not unsafe, they just require a cancellation token 🙂 . Maybe I should remove the `safe_` functions to avoid confusion?

Reply:
I renamed the functions. Hopefully without the "safe_" prefix it is less confusing now.